### ISSUE TITLE
Possible used-before-assignment false positive on type hinting identical to func names in class 

### ISSUE THREAD

**Issue by yaelmi3** [pylint-dev/pylint#9115]
### Bug description

```python
Consider the following class example:

import json
from typing import Optional


class Message:
    message_type: str
    exchange: Optional[str]
    routing_key: str
    payload: dict | str

    def dict(self):
        return {"message_type": self.message_type, "exchange": self.exchange, "payload": self.payload}

    def str(self):
        return json.dumps(self.dict())
```
```


### Configuration

```ini
[MESSAGES CONTROL]
disable=R,missing-docstring,redefined-outer-name

[FORMAT]
max-line-length=120

[REPORTS]
reports=no

[MASTER]
extension-pkg-whitelist=pydantic

[TYPECHECK]
disable=
    no-self-argument

[TYPECHECK.messages control]
generated-members=pydantic.*
```

### Command used

```shell
pylint --rcfile=.pylintrc message.py
```

### Pylint output

```python
************* Module message
cronenberg/message.py:6:18: E0601: Using variable 'str' before assignment (used-before-assignment)
cronenberg/message.py:9:13: E0601: Using variable 'dict' before assignment (used-before-assignment)
```

### Expected behavior

This is a new behaviour encountered in `pylint 3.0.0`, which did not occur in `2.17.7`. pylint treats type hints as possible func names that are defined later on within the class 

### Pylint version

```shell
pylint 3.0.0
astroid 3.0.0
Python 3.10.12 (main, Jun 20 2023, 17:00:24) [Clang 14.0.3 (clang-1403.0.22.14.1)]
```

### OS / Environment

MacOS

**Comment by nickdrozd** [comment]
Simple reproducer:

```python
class A:
    x: str

    def str(self):
        pass
```

**Comment by DanielNoord** [comment]
Note that `mypy` correctly understands that the first example is valid and the second isn't. It should also accept putting the definition of `f` before `str`.
